library/bignum.c

```diff
@@ -2205,23 +2212,23 @@ int mbedtls_mpi_gen_prime( mbedtls_mpi *X, size_t nbits, int dh_flag,
 
     n = BITS_TO_LIMBS( nbits );
 
+new_number:
     MBEDTLS_MPI_CHK( mbedtls_mpi_fill_random( X, n * ciL, f_rng, p_rng ) );
+    /* make sure generated number is at least (n-1)+0.5 bits */
+    if( X->p[n-1] < CEIL_MAXUINT_DIV_SQRT2 ) goto new_number;
 
-    k = mbedtls_mpi_bitlen( X );
-    if( k > nbits ) MBEDTLS_MPI_CHK( mbedtls_mpi_shift_r( X, k - nbits + 1 ) );
-
-    mbedtls_mpi_set_bit( X, nbits-1, 1 );
-
+    k = n * biL;
+    if( k > nbits ) MBEDTLS_MPI_CHK( mbedtls_mpi_shift_r( X, k - nbits ) );
```

Comment:
I prefer to not change the semantics of `mbedtls_mpi_gen_prime`, but to, instead, validate that `X >= 2^(nbits - 0.5)` on the caller side in `mbedtls_rsa_gen_key`, where it can be done by temporarily computing `P^2` resp. `Q^2` in the RSA modulus MPI `N`, and checking that its bitlength is `>= nbits`.

Reply:
> I prefer to not change the semantics of mbedtls_mpi_gen_prime

I did it this way to avoid primality checking (the most expensive part of the operation) of about 40% of random numbers more than ½\*2^n but less than √½\*2^n, which we know are going to be rejected later. I can implement the way you suggest if you're ok with the performance penalty.